compiler/src/iree/compiler/Dialect/Stream/Transforms/ReplicateGlobalsPerAffinity.cpp

```diff
@@ -0,0 +1,189 @@
+}
+
+void ReplicateGlobalsPerAffinityPass::runOnOperation() {
+  mlir::ModuleOp moduleOp = getOperation();
+  IREE::Stream::AffinityAnalysis affinityAnalysis(moduleOp);
+  if (failed(affinityAnalysis.run())) {
+    LDBG() << "failed on running affinity analysis";
+    return;
+  }
+
+  Explorer explorer(moduleOp.getOperation(), TraversalAction::RECURSE);
+  explorer.setOpInterfaceAction<mlir::FunctionOpInterface>(
+      TraversalAction::RECURSE);
+  explorer.setDialectAction<mlir::scf::SCFDialect>(TraversalAction::RECURSE);
+  explorer.setDialectAction<IREE::Flow::FlowDialect>(TraversalAction::RECURSE);
+  explorer.setDialectAction<IREE::Util::UtilDialect>(TraversalAction::RECURSE);
+  explorer.initialize();
+
+  SmallVector<IREE::Util::GlobalOpInterface> candidates;
+  explorer.forEachGlobal([&](const Explorer::GlobalInfo *globalInfo) {
+    if (globalInfo->isIndirect || globalInfo->op.isGlobalMutable()) {
+      return;
+    }
+    candidates.push_back(globalInfo->op);
+  });
+
+  // List of operands that are from global ops:
+  //   [dispatch op, [(operand number, global op)]].
+  DenseMap<IREE::Flow::DispatchOp,
+           SetVector<std::tuple<OpOperand *, IREE::Util::GlobalOpInterface>>>
+      dispatchOpToGlobalUseMap;
+  for (IREE::Util::GlobalOpInterface globalOp : candidates) {
+    const Explorer::GlobalInfo *globalInfo = explorer.getGlobalInfo(globalOp);
+    for (auto loadOp : globalInfo->getLoads()) {
+      explorer.walkTransitiveUses(
+          loadOp.getLoadedGlobalValue(), [&](OpOperand &operand) {
+            TypeSwitch<Operation *>(operand.getOwner())
+                .Case<IREE::Flow::DispatchOp>([&](auto dispatchOp) {
+                  dispatchOpToGlobalUseMap[dispatchOp].insert(
+                      {&operand, globalOp});
+                })
+                .Default([&](Operation *) {});
+            return WalkResult::advance();
+          });
+    }
+  }
+
+  // Map from global ops to affinities. Each global op will be cloned per
+  // affinity.
+  DenseMap<IREE::Util::GlobalOpInterface, SetVector<IREE::Stream::AffinityAttr>>
+      globalAffinityUseMap;
+  // Map from global ops to dispatch ops that use the global op, and the
+  // affinity that the dispatch op runs on.
+  using AffinityUpdatePair =
+      std::tuple<OpOperand *, IREE::Stream::AffinityAttr>;
+  DenseMap<IREE::Util::GlobalOpInterface, SmallVector<AffinityUpdatePair>>
+      globalToDispatchUsesMap;
+  for (auto [dispatchOp, globalUses] : dispatchOpToGlobalUseMap) {
+    // Filter the global operands when we infer the execution affinity for the
+    // dispatch op.
+    LDBG() << "processing dispatch op: " << dispatchOp;
+    SetVector<int> operandFromGlobals;
+    for (auto [operand, globalOp] : globalUses) {
+      (void)globalOp;
+      operandFromGlobals.insert(operand->getOperandNumber());
+    }
+    llvm::SmallSetVector<AffinityAttr, 4> affinityAttrs;
+    for (auto [idx, operand] : llvm::enumerate(dispatchOp.getArguments())) {
```

Comment:
This is a bug because dispatchOp does not only have arguments. It should iterate OpOperands.